src/odemis/gui/comp/overlay/shapes.py

```diff
@@ -92,7 +99,41 @@ def move_to(self, pos: Union[Tuple[float, float], Vec]):
 
     @abstractmethod
     def refresh(self):
-        """Refresh the shape's points VA after copy or move_to operation."""
+        """Refresh the shape's points VA after copy, move_to or restore_state operation."""
+        pass
+
+    def undo(self):
+        """Undo the last action."""
+        if self.undo_stack:
+            state = self.undo_stack.pop()
+            self.redo_stack.append(state)
+            self.restore_state(state)
+            self.cnvs.request_drawing_update()
+        else:
+            logging.warning(
+                "No undo action for %s at %s.", self.__class__.__name__, hex(id(self))
+            )
+
+    def redo(self):
+        """Redo the last undone action."""
+        if self.redo_stack:
+            state = self.redo_stack.pop()
+            self.undo_stack.append(state)
+            self.restore_state(state)
+            self.cnvs.request_drawing_update()
+        else:
+            logging.warning(
+                "No redo action for %s at %s.", self.__class__.__name__, hex(id(self))
+            )
+
+    @abstractmethod
+    def save_state(self):
+        """Save the current state of the shape."""
+        pass
```

Comment:
I think you should be able to simplify a little bit the code by having a simpler `get_state` function that just returns the state of the shape. Then `save_state()` could be generic here, calling `get_state()` and appending it to the the redo stack (with a depth limit).

It would also allow to simplify the .copy(s) function to just `self.restore_state(s.get_state())`.